### PR TITLE
Support borrowed ranges

### PR DESCRIPTION
**Pull request by sylveon**
This marks C++/WinRT ranges as "borrowed": they can be taken by value and iterators to them can be returned without dangling. It's safe for C++/WinRT because collections are reference types, so the lifetime issues typical of C++ collections do not apply.

This enables range algorithms that would previously return `std::ranges::dangling` (a dummy object that isn't dereferencable) to return the appropriate iterator instead. For example, `std::ranges::max_element` or `std::ranges::find` would be quite useless without this, as you can't do anything useful without the returned iterator.

Added tests confirming this works, as well.

The inner range namespace in the codegen is a workaround for [a compiler bug](https://developercommunity.visualstudio.com/t/symbol-cannot-be-defined-within-namespa/1650563?entry=myfeedback).

### CODE REVIEW

File: strings/base_includes.h [hunk 1]
```diff
@@ -28,6 +28,10 @@
 #include <format>
 #endif
 
+#ifdef __cpp_lib_ranges
```
Comment (CaseyCarter) [inline strings/base_includes.h:31]: Pre-existing: I suggest "pre-loading" the feature-test macros with:
```c++
#ifdef __has_include
#if __has_include(<version>)
#include <version>
#endif
#endif
```
before testing them like this. (MSVC's STL defines all feature-test macros when any STL header is included, but that behavior is not required by the C++ Standard and hence not portable.)

File: strings/base_iterator.h [hunk 2]
```diff
@@ -126,7 +126,7 @@ namespace winrt::impl
```
Comment (CaseyCarter) [inline strings/base_iterator.h:126]: Pre-existing: `n - it` (the `operator-(difference_type n, fast_iterator it)` defined on lines 122-125 above) is completely bogus, and should be struck.

File: test/test_cpp20/ranges.cpp [hunk 3]
```diff
@@ -45,4 +45,20 @@ TEST_CASE("ranges")
 
         REQUIRE((result == std::vector{ 2, 4, 6 }));
     }
+
+    // borrowable ranges. without enabling them, *result would fail to compile.
```
Comment (CaseyCarter) [inline test/test_cpp20/ranges.cpp:49]: "Without <feature> this test would not compile" is pretty generic. (Nit: the standard terminology is actually "borrowed range" as made concrete by the `borrowed_range` concept, rather than "borrowable range".)

Maybe just "Validate that rvalue winrt collections are borrowed ranges"?

File: strings/base_iterator.h (hunk 2 above)
Comment (sylveon) [reply strings/base_iterator.h:126]: While you're right, I fear at this point it would be an API break. So I'll leave the final call to @kennykerr.

File: strings/base_includes.h (hunk 1 above)
Comment (sylveon) [reply strings/base_includes.h:31]: Good idea, will add

File: test/test_cpp20/ranges.cpp (hunk 3 above)
Comment (sylveon) [reply test/test_cpp20/ranges.cpp:49]: Borrow is getting pretty overloaded these days haha. Will also change this.

File: strings/base_iterator.h (hunk 2 above)
Comment (kennykerr) [reply strings/base_iterator.h:126]: If Casey says its bogus, I think we should probably fix/remove it. Just include a test for whatever the correct behavior should be, if any.

**Comment by sylveon** [comment]
I added the <version> include, removed the bogus operator on fast_iterator and updated comments in my test!

Had to mess around with the NuGet unit tests to get a full test run passing (they wouldn't build on VS 2022 before).

**Review by CaseyCarter — APPROVED** [review]

File: strings/base_iterator.h [hunk 4]
```diff
@@ -126,7 +126,7 @@ namespace winrt::impl
 
     private:
 
-        T const* m_collection = nullptr;
```
Comment (kennykerr) [inline strings/base_iterator.h:129]: This change would add a `AddRef`/`Release` bump to each iteration, which is why it's always been a reference. Is this strictly necessary?

Comment (sylveon) [reply strings/base_iterator.h:129]: We need `fast_iterator` to hold a strong reference to the collection otherwise it would be dangling when returned and therefore cannot be borrowed (I initially forgot to do that and the unit tests found the issue right away because it crashed).

File: cppwinrt/code_writers.h [hunk 5]
```diff
@@ -158,8 +158,15 @@ namespace cppwinrt
         }
     }
 
+    template<std::size_t indents = 0>
```
Comment (kennykerr) [inline cppwinrt/code_writers.h:161]: Rather than nesting namespaces, just use `namespace std::ranges` in a distinct top-level code gen block. That would be consistent with how the existing namespace are generated.

File: cppwinrt/helpers.h [hunk 6]
```diff
@@ -502,8 +502,13 @@ namespace cppwinrt
...
         {
-            const auto iface = impl.Interface();
-            if (iface.type() != TypeDefOrRef::TypeSpec && type_name(iface) == name)
+            auto iface = impl.Interface();
+            if (iface.type() == TypeDefOrRef::TypeSpec)
+            {
+                iface = iface.TypeSpec().Signature().GenericTypeInst().GenericType();
+            }
+
+            if (type_name(iface) == name)
```
Comment (kennykerr) [inline cppwinrt/helpers.h:511]: Can you explain why this change was needed?

File: test/test/fast_iterator.cpp [hunk 7]
```diff
@@ -45,7 +45,7 @@ TEST_CASE("fast_iterator")
         REQUIRE(vbegin[2] == 4);
         REQUIRE(vbegin + 2 > vbegin);
         REQUIRE(2 + vbegin > vbegin);
-        REQUIRE(2 - (vbegin + 4) > vbegin);
+        REQUIRE(vbegin + 4 - 2 > vbegin);
```
Comment (kennykerr) [inline test/test/fast_iterator.cpp:48]: Did this break?

File: strings/base_iterator.h (hunk 4 above)
Comment (kennykerr) [reply strings/base_iterator.h:129]: But that means folks using it in a range-base for loop are paying the price (the scenario for which the `fast_iterator` was originally conceived).

Comment (kennykerr) [reply strings/base_iterator.h:129]: Although not expensive, just as you wouldn't expect to take a copy of a `std::vector` to make an iterator, you shouldn't make a copy of the collection interface either.

File: test/test/fast_iterator.cpp (hunk 7 above)
Comment (CaseyCarter) [reply test/test/fast_iterator.cpp:48]: Yes, this was testing the bogus "integer - iterator" operation. It probably makes more sense to delete the test case than to change it into a redundant test of other operations.

**Review by kennykerr — CHANGES_REQUESTED** [review]
Let's figure out how to do this without costing more for existing code.

File: strings/base_iterator.h (hunk 4 above)
Comment (sylveon) [reply strings/base_iterator.h:129]: IIRC, range-based for loops use `++i`, not `i++`, so they don't pay the cost.

File: cppwinrt/helpers.h (hunk 6 above)
Comment (sylveon) [reply cppwinrt/helpers.h:511]: IIterable is a generic interface, and the previous code skipped over them

File: test/test/fast_iterator.cpp (hunk 7 above)
Comment (sylveon) [reply test/test/fast_iterator.cpp:48]: Yes, it was the bogus version. It's already covered down below (`vbegin + 2 - 2`), so removing this line makes sense

File: strings/base_iterator.h (hunk 4 above)
Comment (kennykerr) [reply strings/base_iterator.h:129]: I'm talking about the cost of creating the `fast_iterator` object.

Comment (sylveon) [reply strings/base_iterator.h:129]: While it does technically get more costly, I feel like this one-time cost to be pretty minimal compared to the actual loop body, especially since WinRT is involved: every iteration of the ranged-for loop will do a AddRef/Release too because of GetAt. Not to mention that getting any property on the result of GetAt, or copying it elsewhere, etc. will also involve AddRef/Release.

A partial amount of the cost is solved by compiler-autogenerated move constructors/assignment operator.

The change also helps consistency between `fast_iterator` and `IIterator<T>` in a scenario like the following code.
```cpp
auto v = winrt::single_threaded_vector<int>({1, 2, 3});
auto it = begin(v);
v = somethingElse;
*it;
```
The `*it` expression will get the first value of whatever `somethingElse` contains. because `fast_iterator` holds a pointer to the variable `v` and `v` was reassigned after calling `begin(v)`.

If we compare against the same code, but boxing to `IIterable<int>` first:
```cpp
IIterable<int> v = winrt::single_threaded_vector<int>({1, 2, 3});
auto it = begin(v);
v = somethingElse;
*it;
```
Now `*it` will get the first value from the vector that was created on line 1, because `IIterator<int>` holds a reference to the source vector. With this change, `fast_iterator` behaves the same as `IIterator<T>`

Comment (kennykerr) [reply strings/base_iterator.h:129]: > every iteration of the ranged-for loop will do a AddRef/Release too because of GetAt

There's no `AddRef`/`Release` bump when calling `GetAt` on a collection of `int` as you've used in this example. The only cost is a single virtual function call, `GetAt` itself. Yes, if your loop body is expensive it may not matter that much. On the other hand, if you're just iterator a collection of `int` then it may be impactful. I'm not trying to be difficult, but I have customers that really really care about getting rid of every last unnecessary `AddRef`/`Release` pair. 

I'm not familiar with C++ ranges - did `std::vector` iterators suddenly get more expensive to accommodate ranges? That would be surprising.

Comment (sylveon) [reply strings/base_iterator.h:129]: No, but for example in debug mode it's not cheap to build one, as iterator debugging takes a lock iirc. (debug iterators don't have a trivial destructor either).

We could experiment with overloading begin for rvalue refs (e.g. `auto begin() &&;`) to return an owning iterator (moving ownership into it) and leaving the normal begin to return a non-owning iterator, eliding an `AddRef`/`Release` for the case of borrowed ranges, and leaving for loops untouched. We would need to add `bool owning;` to `fast_iterator` as ranges wants the same iterator type everywhere, and `iterator_t` uses an lvalue ref to determine the iterator type, so returning a different type from our overload is not possible. That would certainly complexify the implementation of `fast_iterator`.

Or we could disable borrowed ranges for `IVector` and `IBindableVector` (forcing explicit ownership, or converting to `IIterable` and losing out on `fast_iterator`), but that will lead to inconsistency in the API that I'm not a huge fan of.

Obviously the easiest way is just staying as is and dealing with the additional `AddRef` (using WinRT in a performance critical loop is already not the best thing), but that's already been disqualified.

Comment (kennykerr) [reply strings/base_iterator.h:129]: Sounds even less appealing. 😉 Why don't we go back to the beginning and deal with the question of why this is being done in the first place. There was no issue opened prior to this PR so I have no context. What fundamental problem in C++/WinRT does this solve? Why do we need to "support borrowed ranges"? Why the extra complexity and cost?

Comment (sylveon) [reply strings/base_iterator.h:129]: Supporting borrowed ranges allows consumers to directly pass the return value of an API call to the `std::ranges` algorithms without needing to store it in a variable first, and the return value will still be usable. It basically signals to those algorithms that the type it's iterating is either non-owning (`std::span` and `std::string_view` for example), or has iterators that holds a strong reference to the underlying collection (WinRT collection types). It allows us to simplify code.

```cpp
PackageManager manager;
IIterator<Package> someAppPackage = std::ranges::find_if(manager.FindPackagesForUser(L""),
    [](const Package &package)
    {
        return package.Id().FullName() == L"SomeApp_8wekyb3d8bbwe";
    });

if (someAppPackage != nullptr)
{
    // valid and not dangling
    (*someAppPackage).Launch(L"");
}
```

Without specializing `enabled_borrowed_range` to say `true`, `std::ranges::find_if` in this example above would return `std::ranges::dangling`, which is a fake object that disallows any kind of useful operation, because it assumes the iterator does not hold a strong reference so the collection would be destroyed once `std::ranges:find_if` returns and the iterator would be dangling.

Comment (kennykerr) [reply strings/base_iterator.h:129]: You can write that more simply without ranges. 

```C++
PackageManager manager;
for (auto package : manager.FindPackagesForUser(L""))
{
    if (package.Id().FullName() == L"SomeApp_8wekyb3d8bbwe")
    {
        package.Launch(L"");
    }
}
```

Comment (sylveon) [reply strings/base_iterator.h:129]: True, albeit that was an example. There are obviously more complex algorithms that would be harder to write and/or maintain as a normal for loop.

Comment (kennykerr) [reply strings/base_iterator.h:129]: Understood, it sounds like it really needs a larger investment to get ranges supported without compromising performance. We've spoken about making C++20 the minver in future. This sounds like a good feature to include in such a rejuvenation effort.
